### PR TITLE
Fix crash in Macro.to_string

### PR DESCRIPTION
**Pull request by sabiwara**
This is the first issue described in https://github.com/elixir-lang/elixir/issues/13895.

### CODE REVIEW

File: lib/elixir/lib/code/formatter.ex [hunk 1]
```diff
@@ -1362,43 +1362,57 @@ defmodule Code.Formatter do
...
+
     interpolation_to_algebra(entries, escape, state, concat(acc, doc), last)
   end
 
   defp interpolation_to_algebra([], _escape, state, acc, last) do
     {concat(acc, last), state}
   end
 
-  defp interpolation_to_algebra(quoted, %{skip_eol: skip_eol} = state) do
+  defp interpolated_to_algebra(quoted, %{skip_eol: skip_eol} = state) do
```
Comment (sabiwara) [inline lib/elixir/lib/code/formatter.ex:1399]: I renamed this because I found it confusing, but if anybody has another suggestion I'm happy to take it.

**Comment by josevalim** [comment]
I think the bug is actually in the normalizer. :( We only convert the sigil to proper syntax if it has a delimiter:

https://github.com/elixir-lang/elixir/blob/3ce07dc743f3879345e05c500b58d76e3383b80e/lib/elixir/lib/code/formatter.ex#L1421

I think the issue is that the normalizer is adding a delimiter, when it is not a proper AST. WDYT?

**Comment by sabiwara** [comment]
OK, so the idea is to treat `sigil_r(<<"abc", 123>>, [])` just as if it was not a sigil, same as `foo(<<"abc", 123>>, [])`?

Currently there are two differences:
- adding a delimiter
- wrapping literal args

```elixir
iex> Code.Normalizer.normalize(quote do: foo(<<"abc", 123>>, []))
{:foo, [closing: [line: nil]],
 [
   {:<<>>, [], [{:__block__, [], ["abc"]}, {:__block__, [token: "123"], ~c"{"}]},
   {:__block__, [], [[]]}  # <=== [] is wrapped
 ]}
iex> Code.Normalizer.normalize(quote do: sigil_r(<<"abc", 123>>, []))
{:sigil_r, [delimiter: "\"", context: Elixir, imports: [{2, Kernel}]],
 [
   {:<<>>, [], [{:__block__, [], ["abc"]}, {:__block__, [token: "123"], ~c"{"}]},
   []                     # <=== [] is NOT wrapped
 ]}
```

**Comment by josevalim** [comment]
Exactly. We should not add a delimiter and do other sigil things if it is not a sigil.

**Comment by sabiwara** [comment]
@josevalim makes sense to me, re-pushed with the new approach [adfe040](https://github.com/elixir-lang/elixir/pull/13905/commits/adfe0406079fe86b5eb0149e18ba8c8c8738a9c2).

I wonder though if it could be considered a regression, this felt intentional based on this commit https://github.com/sabiwara/elixir/commit/2d1c64cc5d97cdba1e9f05de88623eafc46dfe3d. Maybe when creating sigils through meta-programming, if that's a thing?

**Comment by josevalim** [comment]
@sabiwara we should definitely support sigils through meta-programming. I think we should not check for delimiters in Code.Normalizer. Instead, we should normalize that both arguments of the sigil macro are "valid" (from the point of view of the formatter) and only add the delimiter if we can normalize them.

**Comment by sabiwara** [comment]
OK I see what you mean now, sorry for the misunderstanding 🙇

BTW, unrelated but [this](https://github.com/elixir-lang/elixir/pull/13905/commits/9dae94c6927507e938cd2bf71a790cc4c2bacc9b) feels like a good change, there is no need to lose the original delimiter in ExUnit no? I won't need it anymore for this PR though.

**Comment by josevalim** [comment]
Agreed it is a good change!

**Comment by sabiwara** [comment]
@josevalim forced-pushed take 3: https://github.com/elixir-lang/elixir/pull/13905/commits/020025aa0a1a068c95d1c73ea1d5e857ae19e319.

I also caught a couple of crashes when the modifiers are invalid, covered these cases too.

**Comment by sabiwara** [comment]
[Backported](https://github.com/elixir-lang/elixir/tree/v1.17)

**Pull request merged**